qml/Components/Wallpaper.qml

```diff
@@ -20,12 +21,76 @@ import Ubuntu.Components 1.3
 
 Item {
     id: root
-    property url source
+    objectName: "Wallpaper"
+    property url source: ""
+
+    // sourceSize should be set to the largest width or height that this
+    // Wallpaper will likely be set to.
+    // In most cases inside Lomiri, this should be set to the width or height
+    // of the shell, whichever is larger.
+    // However, care should be taken to avoid changing sourceSize once
+    // it's set.
+    property real sourceSize: -1.0
 
     Image {
+        id: image
+        objectName: "wallpaperImage"
         anchors.fill: parent
         fillMode: Image.PreserveAspectCrop
 
-        source: root.source
+        asynchronous: true
+
+        property real oldSourceSize
+        property real intermediarySourceSize: {
+            // If minSourceSize is defined by root's parent, it will be 0 until
+            // the binding resolves. Otherwise, it'll be -1.0. Abuse this
+            // behavior to avoid changing sourceSize more than once during
+            // startup
+            if (root.sourceSize === 0.0) {
+                developerSanityTimer.start();
+                // -1.0 signals that the wallpaper shouldn't render yet.
+                return -1.0
+            }
+            developerSanityTimer.stop();
+            if (root.sourceSize === -1.0 && root.source !== "") {
+                console.warn(`${root.objectName}'s sourceSize is unset. It will use a lot of memory.`);
+                return 0.0
+            }
+
+            return root.sourceSize
+        }
+
+        onIntermediarySourceSizeChanged: {
+            if (oldSourceSize !== 0.0 && oldSourceSize !== -1.0 && root.source !== "") {
+                console.warn(`${root.objectName} source size is changing from ${oldSourceSize} to ${intermediarySourceSize}. This will cause an expensive image reload.`);
+            }
+            oldSourceSize = intermediarySourceSize;
+        }
+
+        sourceSize: Qt.size(intermediarySourceSize, intermediarySourceSize)
+        source: intermediarySourceSize !== -1.0 ? root.source : ""
+    }
+
+    Rectangle {
+        id: wallpaperFadeRectangle
+        objectName: "wallpaperFadeRectangle"
+        color: theme.palette.normal.background
+        anchors.fill: parent
+        opacity: image.status === Image.Ready ? 0: 1
+        visible: opacity !== 0
+        Behavior on opacity {
+            UbuntuNumberAnimation { duration: UbuntuAnimation.FastDuration }
+        }
+    }
+
+    Timer {
```

Comment:
If you used a delayed binding for the source property, you could get rid of this.

Reply:
I'm not sure I understand. What am I replacing with a delayed binding?

Reply:
I meant that instead of using a Timer here, you could have the `source` changed signal connected with a delayed binding, and at that point in the handler of this binding you could be sure that the sourceSize must have been set. If it isn't set, then you'd print the error message. But it's a minor thing, so do it only if you like: I'm approving this anyways.